File: packages/vscode-extension/src/chat/ide-session-index.js
/**
 * Shared IDE session index.
 *
 * VS Code and JetBrains both write ~/.chainlesschain/ide/session-index.json so
 * a session created in one IDE is discoverable from the other's session picker.
 * This deliberately stores metadata only; transcripts stay in the CLI session
 * store and are resumed by id.
 */
const fs = require("fs");
const os = require("os");
const path = require("path");

const VALID_STATUS = new Set([
  "running",
  "waiting_approval",
  "errored",
  "stopped",
  "completed",
]);
const MAX_RECORDS = 200;
const DEFAULT_LOCK_TIMEOUT_MS = 2000;
const DEFAULT_LOCK_STALE_MS = 30000;
const DEFAULT_LOCK_RETRY_MS = 25;

function defaultIndexFile(home = os.homedir()) {
  return path.join(home, ".chainlesschain", "ide", "session-index.json");
}

function cleanString(v) {
  return typeof v === "string" ? v.trim() : "";
}

function cleanStatus(v) {
  const s = cleanString(v).replace(/\s+/g, "_").toLowerCase();
  return VALID_STATUS.has(s) ? s : "stopped";
}

function iso(v, fallback) {
  if (v instanceof Date && !Number.isNaN(v.getTime())) return v.toISOString();
  if (typeof v === "string" && v) {
    const d = new Date(v);
    if (!Number.isNaN(d.getTime())) return d.toISOString();
  }
  return fallback;
}

function normalizeRecord(input = {}, { now = new Date() } = {}) {
  const id = cleanString(input.id || input.sessionId);
  if (!id) return null;
  const updatedAt = iso(input.updatedAt, now.toISOString());
  const createdAt = iso(input.createdAt, updatedAt);
  const folders = Array.isArray(input.workspaceFolders)
    ? input.workspaceFolders.map(cleanString).filter(Boolean).slice(0, 8)
    : [];
  const workspace = cleanString(input.workspace) || folders[0] || "";
  return {
    id,
    title: cleanString(input.title),
    ide: cleanString(input.ide) || "unknown",
    conversationId: cleanString(input.conversationId),
    workspace,
    workspaceFolders: folders,
    status: cleanStatus(input.status),
    mode: cleanString(input.mode) || "default",
    createdAt,
    updatedAt,
  };
}

function indexError(code, message, cause) {
  const error = new Error(message, cause ? { cause } : undefined);
  error.code = code;
  return error;
}

function parseIndex(raw, { failIfUnavailable = false } = {}) {
  let parsed;
  try {
    parsed = JSON.parse(String(raw || "").trim());
  } catch (cause) {
    if (failIfUnavailable) {
      throw indexError(
        "IDE_SESSION_INDEX_CORRUPT",
        "IDE session index contains invalid JSON",
        cause,
      );
    }
    return [];
  }
  const rows = Array.isArray(parsed)
    ? parsed
    : Array.isArray(parsed?.sessions)
      ? parsed.sessions
      : [];
  if (
    failIfUnavailable &&
    !Array.isArray(parsed) &&
    !Array.isArray(parsed?.sessions)
  ) {
    throw indexError(
      "IDE_SESSION_INDEX_CORRUPT",
      "IDE session index must contain a sessions array",
    );
  }
  return rows.map((r) => normalizeRecord(r)).filter(Boolean);
}

function mergeRecords(existing, incoming, { limit = MAX_RECORDS, now } = {}) {
  const byId = new Map();
  for (const row of existing || []) {
    const rec = normalizeRecord(row, { now });
    if (rec) byId.set(rec.id, rec);
  }
  for (const row of Array.isArray(incoming) ? incoming : [incoming]) {
    const rec = normalizeRecord(row, { now });
    if (!rec) continue;
    const prev = byId.get(rec.id);
    byId.set(rec.id, {
      ...(prev || {}),
      ...rec,
      createdAt: prev?.createdAt || rec.createdAt,
      title: rec.title || prev?.title || "",
      workspace: rec.workspace || prev?.workspace || "",
      workspaceFolders: rec.workspaceFolders.length
        ? rec.workspaceFolders
        : prev?.workspaceFolders || [],
    });
  }
  return [...byId.values()]
    .sort((a, b) => String(b.updatedAt).localeCompare(String(a.updatedAt)))
    .slice(0, limit);
}

function readIdeSessionIndex({
  file = defaultIndexFile(),
  deps = fs,
  failIfUnavailable = false,
} = {}) {
  try {
    return parseIndex(deps.readFileSync(file, "utf8"), {
      failIfUnavailable,
    });
  } catch (cause) {
    if (cause?.code === "ENOENT") return [];
    if (failIfUnavailable || cause?.code === "IDE_SESSION_INDEX_CORRUPT") {
      throw cause;
    }
    return [];
  }
}

function sleepSync(ms) {
  try {
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
  } catch {
    const deadline = Date.now() + ms;
    while (Date.now() < deadline) {
      // SharedArrayBuffer is unavailable; retain a bounded fallback.
    }
  }
}

/**
 * Cross-process protocol shared with the JetBrains implementation. The lock is
 * an atomic sibling DIRECTORY (`session-index.json.lock`), not an advisory file
 * lock, so Node and JVM writers exclude one another on Windows and POSIX.
 */
function withIdeSessionIndexLock(file, body, opts = {}) {
  const deps = opts.deps || fs;
  const timeoutMs = opts.lockTimeoutMs ?? DEFAULT_LOCK_TIMEOUT_MS;
  const staleMs = opts.lockStaleMs ?? DEFAULT_LOCK_STALE_MS;
  const retryMs = opts.lockRetryMs ?? DEFAULT_LOCK_RETRY_MS;
  const nowMs = opts.nowMs || Date.now;
  const sleep = opts.sleep || sleepSync;
  const lockDir = `${file}.lock`;
  const deadline = nowMs() + timeoutMs;
  let held = false;

  try {
    deps.mkdirSync(path.dirname(file), { recursive: true });
  } catch (cause) {
    throw indexError(
      "IDE_SESSION_INDEX_PREPARE_FAILED",
      `Could not prepare IDE session index directory: ${path.dirname(file)}`,
      cause,
    );
  }

  for (;;) {
    try {
      deps.mkdirSync(lockDir);
      held = true;
      break;
    } catch (cause) {
      if (cause?.code !== "EEXIST") {
        throw indexError(
          "IDE_SESSION_INDEX_LOCK_UNAVAILABLE",
          `Could not acquire IDE session index lock: ${lockDir}`,
          cause,
        );
      }
      try {
        const age = nowMs() - deps.statSync(lockDir).mtimeMs;
        if (age > staleMs) {
          deps.rmSync(lockDir, { recursive: true, force: true });
          continue;
        }
      } catch {
        // The holder may have released between mkdir and stat; retry boundedly.
      }
      if (nowMs() >= deadline) {
        throw indexError(
          "IDE_SESSION_INDEX_LOCK_UNAVAILABLE",
          `Timed out acquiring IDE session index lock: ${lockDir}`,
        );
      }
      sleep(retryMs);
    }
  }

  try {
    return body();
  } finally {
    if (held) {
      try {
        deps.rmSync(lockDir, { recursive: true, force: true });
      } catch {
        // The mutation is already durable. A stale lock is reclaimed later.
      }
    }
  }
}

function writeIdeSessionIndexUnlocked(
  records,
  { file = defaultIndexFile(), deps = fs } = {},
) {
  const dir = path.dirname(file);
  deps.mkdirSync(dir, { recursive: true });
  const body = JSON.stringify({ version: 1, sessions: records || [] }, null, 2);
  const tmp = `${file}.${process.pid}.${Date.now()}.tmp`;
  try {
    deps.writeFileSync(tmp, body, { encoding: "utf8", mode: 0o600 });
    try {
      deps.chmodSync(tmp, 0o600);
    } catch {
      /* best-effort */
    }
    deps.renameSync(tmp, file);
  } catch (cause) {
    try {
      deps.rmSync(tmp, { force: true });
    } catch {
      // Preserve the original persistence error.
    }
    throw indexError(
      "IDE_SESSION_INDEX_WRITE_FAILED",
      `Could not atomically write IDE session index: ${file}`,
      cause,
    );
  }
  return file;
}

function writeIdeSessionIndex(records, opts = {}) {
  const file = opts.file || defaultIndexFile();
  return withIdeSessionIndexLock(
    file,
    () => writeIdeSessionIndexUnlocked(records, { ...opts, file }),
    opts,
  );
}

function upsertIdeSessionRecord(record, opts = {}) {
  const now = opts.now instanceof Date ? opts.now : new Date();
  const file = opts.file || defaultIndexFile();
  return withIdeSessionIndexLock(
    file,
    () => {
      const current = readIdeSessionIndex({
        ...opts,
        file,
        failIfUnavailable: true,
      });
      const sessions = mergeRecords(current, record, { now });
      writeIdeSessionIndexUnlocked(sessions, { ...opts, file });
      return normalizeRecord(record, { now });
    },
    opts,
  );
}

/**
 * Rename a session in the shared index. For sessions the index has never seen
 * (CLI-only sessions) a minimal overlay record is created — the picker merge
 * prefers the IDE-index title, so the rename shows up for those too.
 */
function renameIdeSessionRecord(id, title, opts = {}) {
  const key = cleanString(id);
  const nextTitle = cleanString(title);
  if (!key || !nextTitle) return null;
  const now = opts.now instanceof Date ? opts.now : new Date();
  const file = opts.file || defaultIndexFile();
  return withIdeSessionIndexLock(
    file,
    () => {
      const current = readIdeSessionIndex({
        ...opts,
        file,
        failIfUnavailable: true,
      });
      const existing = current.find((r) => r.id === key);
      const record = {
        ...(existing || { id: key }),
        title: nextTitle,
        updatedAt: now.toISOString(),
      };
      writeIdeSessionIndexUnlocked(mergeRecords(current, record, { now }), {
        ...opts,
        file,
      });
      return normalizeRecord(record, { now });
    },
    opts,
  );
}

/** Drop a session from the shared index. Returns whether anything was removed. */
function removeIdeSessionRecord(id, opts = {}) {
  const key = cleanString(id);
  if (!key) return false;
  const file = opts.file || defaultIndexFile();
  return withIdeSessionIndexLock(
    file,
    () => {
      const current = readIdeSessionIndex({
        ...opts,
        file,
        failIfUnavailable: true,
      });
      const next = current.filter((r) => r.id !== key);
      if (next.length === current.length) return false;
      writeIdeSessionIndexUnlocked(next, { ...opts, file });
      return true;
    },
    opts,
  );
}

function toSessionItems(records) {
  return (records || []).map((s) => ({
    id: s.id,
    title: s.title || "",
    updatedAt: s.updatedAt || null,
    store: `ide:${s.ide || "unknown"}`,
    status: s.status || "stopped",
    workspace: s.workspace || "",
  }));
}

module.exports = {
  defaultIndexFile,
  mergeRecords,
  normalizeRecord,
  parseIndex,
  readIdeSessionIndex,
  removeIdeSessionRecord,
  renameIdeSessionRecord,
  toSessionItems,
  upsertIdeSessionRecord,
  withIdeSessionIndexLock,
  writeIdeSessionIndex,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/IdeSessionIndex.java
package com.chainlesschain.ide;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Shared IDE session index.
 *
 * VS Code and JetBrains both write ~/.chainlesschain/ide/session-index.json so
 * either IDE can discover and resume sessions created by the other. Only
 * metadata is stored here; the transcript remains in the CLI session store.
 */
public final class IdeSessionIndex {
    private IdeSessionIndex() {}

    public static final int MAX_RECORDS = 200;

    // Every mutation is a read-merge-write of the whole file; two writers
    // interleaving lose one of the updates. In-process: one monitor. Cross-
    // process (VS Code writes the same file): an OS lock on a sibling .lock
    // file — best-effort, a lock failure NEVER drops the write (fail open;
    // losing exclusion beats losing the upsert).
    private static final Object IO_LOCK = new Object();
    private static final java.util.concurrent.atomic.AtomicLong TMP_SEQ =
            new java.util.concurrent.atomic.AtomicLong();

    private interface IoMutation {
        void run() throws IOException;
    }

    private static void locked(Path file, IoMutation body) throws IOException {
        synchronized (IO_LOCK) {
            java.nio.channels.FileChannel ch = null;
            java.nio.channels.FileLock lock = null;
            try {
                Files.createDirectories(file.getParent());
                ch = java.nio.channels.FileChannel.open(
                        file.resolveSibling(file.getFileName().toString() + ".lock"),
                        java.nio.file.StandardOpenOption.CREATE,
                        java.nio.file.StandardOpenOption.WRITE);
                lock = ch.lock();
            } catch (Exception ignored) {
                // fail open — cross-process exclusion is best-effort
            }
            try {
                body.run();
            } finally {
                try {
                    if (lock != null) lock.release();
                } catch (Exception ignored) {
                    // already released / channel closed
                }
                try {
                    if (ch != null) ch.close();
                } catch (Exception ignored) {
                    // already closed
                }
            }
        }
    }

    public static Path defaultFile() {
        return Paths.get(System.getProperty("user.home"),
                ".chainlesschain", "ide", "session-index.json");
    }

    public static Map<String, Object> record(String id, String title, String ide,
            String conversationId, String workspace, List<String> workspaceFolders,
            String status, String mode, Instant updatedAt) {
        Map<String, Object> r = new LinkedHashMap<String, Object>();
        r.put("id", clean(id));
        r.put("title", clean(title));
        r.put("ide", clean(ide).isEmpty() ? "unknown" : clean(ide));
        r.put("conversationId", clean(conversationId));
        r.put("workspace", clean(workspace));
        r.put("workspaceFolders", workspaceFolders == null
                ? new ArrayList<String>() : new ArrayList<String>(workspaceFolders));
        r.put("status", cleanStatus(status));
        r.put("mode", clean(mode).isEmpty() ? "default" : clean(mode));
        r.put("updatedAt", String.valueOf(updatedAt != null ? updatedAt : Instant.now()));
        return r;
    }

    public static boolean upsertDefault(Map<String, Object> record) {
        try {
            upsert(defaultFile(), record);
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    public static void upsert(Path file, Map<String, Object> record) throws IOException {
        locked(file, () -> {
            List<Map<String, Object>> merged =
                    merge(read(file), record, Instant.now(), MAX_RECORDS);
            write(file, merged);
        });
    }

    public static boolean renameDefault(String id, String title) {
        try {
            return rename(defaultFile(), id, title);
        } catch (IOException e) {
            return false;
        }
    }

    /**
     * Rename a session in the shared index. For sessions the index has never
     * seen (CLI-only sessions) a minimal overlay record is created — the picker
     * merge prefers the IDE-index title, so the rename shows up for those too.
     */
    public static boolean rename(Path file, String id, String title) throws IOException {
        String key = clean(id);
        String next = clean(title);
        if (key.isEmpty() || next.isEmpty()) return false;
        locked(file, () -> {
            List<Map<String, Object>> rows = read(file);
            Map<String, Object> match = null;
            for (Map<String, Object> row : rows) {
                if (key.equals(row.get("id"))) match = row;
            }
            if (match == null) {
                match = new LinkedHashMap<String, Object>();
                match.put("id", key);
                match = normalize(match, Instant.now());
                rows.add(match);
            }
            match.put("title", next);
            match.put("updatedAt", String.valueOf(Instant.now()));
            write(file, rows);
        });
        return true;
    }

    public static boolean removeDefault(String id) {
        try {
            return remove(defaultFile(), id);
        } catch (IOException e) {
            return false;
        }
    }

    /** Drop a session from the shared index. Returns whether anything was removed. */
    public static boolean remove(Path file, String id) throws IOException {
        String key = clean(id);
        if (key.isEmpty()) return false;
        boolean[] removed = new boolean[1];
        locked(file, () -> {
            List<Map<String, Object>> rows = read(file);
            for (java.util.Iterator<Map<String, Object>> it = rows.iterator(); it.hasNext();) {
                if (key.equals(it.next().get("id"))) {
                    it.remove();
                    removed[0] = true;
                }
            }
            if (removed[0]) write(file, rows);
        });
        return removed[0];
    }

    @SuppressWarnings("unchecked")
    public static List<Map<String, Object>> read(Path file) {
        List<Map<String, Object>> out = new ArrayList<Map<String, Object>>();
        String raw;
        try {
            raw = new String(Files.readAllBytes(file), StandardCharsets.UTF_8).trim();
        } catch (IOException e) {
            return out;
        }
        Object parsed;
        try {
            parsed = MiniJson.parse(raw);
        } catch (RuntimeException e) {
            return out;
        }
        Object rows = parsed instanceof Map ? ((Map<?, ?>) parsed).get("sessions") : parsed;
        if (!(rows instanceof List)) return out;
        for (Object row : (List<?>) rows) {
            if (!(row instanceof Map)) continue;
            Map<String, Object> normalized =
                    normalize((Map<String, Object>) row, Instant.now());
            if (normalized != null) out.add(normalized);
        }
        return out;
    }

    public static List<Map<String, Object>> merge(List<Map<String, Object>> existing,
            Map<String, Object> incoming, Instant now, int limit) {
        Map<String, Map<String, Object>> byId =
                new LinkedHashMap<String, Map<String, Object>>();
        if (existing != null) {
            for (Map<String, Object> row : existing) {
                Map<String, Object> n = normalize(row, now);
                if (n != null) byId.put(String.valueOf(n.get("id")), n);
            }
        }
        Map<String, Object> next = normalize(incoming, now);
        if (next != null) {
            String id = String.valueOf(next.get("id"));
            Map<String, Object> prev = byId.get(id);
            if (prev != null) {
                next.put("createdAt", prev.get("createdAt"));
                if (clean(String.valueOf(next.get("title"))).isEmpty()) {
                    next.put("title", prev.get("title"));
                }
                if (clean(String.valueOf(next.get("workspace"))).isEmpty()) {
                    next.put("workspace", prev.get("workspace"));
                }
            }
            byId.put(id, next);
        }
        List<Map<String, Object>> rows =
                new ArrayList<Map<String, Object>>(byId.values());
        Collections.sort(rows, new Comparator<Map<String, Object>>() {
            public int compare(Map<String, Object> a, Map<String, Object> b) {
                return String.valueOf(b.get("updatedAt"))
                        .compareTo(String.valueOf(a.get("updatedAt")));
            }
        });
        if (rows.size() > limit) return new ArrayList<Map<String, Object>>(rows.subList(0, limit));
        return rows;
    }

    public static void write(Path file, List<Map<String, Object>> sessions) throws IOException {
        Files.createDirectories(file.getParent());
        Map<String, Object> root = new LinkedHashMap<String, Object>();
        root.put("version", 1L);
        root.put("sessions", sessions == null
                ? new ArrayList<Map<String, Object>>() : sessions);
        // Unique per write: two same-millisecond writers used to collide on the
        // tmp name — the second Files.move failed and its upsert vanished.
        Path tmp = file.resolveSibling(file.getFileName().toString()
                + "." + System.nanoTime() + "-" + TMP_SEQ.incrementAndGet() + ".tmp");
        Files.write(tmp, MiniJson.stringify(root).getBytes(StandardCharsets.UTF_8));
        try {
            Files.move(tmp, file, StandardCopyOption.REPLACE_EXISTING,
                    StandardCopyOption.ATOMIC_MOVE);
        } catch (IOException e) {
            Files.move(tmp, file, StandardCopyOption.REPLACE_EXISTING);
        }
    }

    public static List<SessionList.SessionItem> sessionItems(Path file) {
        List<SessionList.SessionItem> out = new ArrayList<SessionList.SessionItem>();
        for (Map<String, Object> row : read(file)) {
            String id = clean(String.valueOf(row.get("id")));
            if (id.isEmpty()) continue;
            String ide = clean(String.valueOf(row.get("ide")));
            out.add(new SessionList.SessionItem(
                    id,
                    clean(String.valueOf(row.get("title"))),
                    clean(String.valueOf(row.get("updatedAt"))),
                    "ide:" + (ide.isEmpty() ? "unknown" : ide),
                    clean(String.valueOf(row.get("status"))),
                    clean(String.valueOf(row.get("workspace")))));
        }
        return out;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> normalize(Map<String, Object> row, Instant now) {
        if (row == null) return null;
        String id = clean(String.valueOf(first(row.get("id"), row.get("sessionId"))));
        if (id.isEmpty()) return null;
        String updatedAt = iso(row.get("updatedAt"), now);
        String createdAt = iso(row.get("createdAt"), Instant.parse(updatedAt));
        List<String> folders = new ArrayList<String>();
        Object rawFolders = row.get("workspaceFolders");
        if (rawFolders instanceof List) {
            for (Object v : (List<Object>) rawFolders) {
                String s = clean(String.valueOf(v));
                if (!s.isEmpty() && folders.size() < 8) folders.add(s);
            }
        }
        String workspace = clean(String.valueOf(row.get("workspace")));
        if (workspace.isEmpty() && !folders.isEmpty()) workspace = folders.get(0);

        Map<String, Object> out = new LinkedHashMap<String, Object>();
        out.put("id", id);
        out.put("title", clean(String.valueOf(row.get("title"))));
        out.put("ide", clean(String.valueOf(row.get("ide"))).isEmpty()
                ? "unknown" : clean(String.valueOf(row.get("ide"))));
        out.put("conversationId", clean(String.valueOf(row.get("conversationId"))));
        out.put("workspace", workspace);
        out.put("workspaceFolders", folders);
        out.put("status", cleanStatus(String.valueOf(row.get("status"))));
        out.put("mode", clean(String.valueOf(row.get("mode"))).isEmpty()
                ? "default" : clean(String.valueOf(row.get("mode"))));
        out.put("createdAt", createdAt);
        out.put("updatedAt", updatedAt);
        return out;
    }

    private static Object first(Object a, Object b) {
        return a != null ? a : b;
    }

    private static String clean(String s) {
        return s == null || "null".equals(s) ? "" : s.trim();
    }

    private static String cleanStatus(String s) {
        String v = clean(s).toLowerCase().replace(' ', '_');
        if ("running".equals(v) || "waiting_approval".equals(v)
                || "errored".equals(v) || "stopped".equals(v)
                || "completed".equals(v)) return v;
        return "stopped";
    }

    private static String iso(Object raw, Instant fallback) {
        if (raw instanceof String) {
            try {
                return String.valueOf(Instant.parse((String) raw));
            } catch (RuntimeException ignored) {
                // fall through
            }
        }
        return String.valueOf(fallback != null ? fallback : Instant.now());
    }
}
